proposals/0072-feature-gate-threshold-automation.md

```diff
@@ -0,0 +1,247 @@
+---
+simd: '0072'
+title: Feature Gate Threshold Automation
+authors:
+  - Tyera Eulburg
+  - Joe Caulfield
+category: Standard
+type: Core
+status: Draft
+created: 2023-10-11
+feature: (fill in with feature tracking issues once accepted)
+---
+
+## Summary
+
+This SIMD outlines a proposal for automating the feature activation process
+based on a stake-weighted support threshold, rather than manual human action.
+
+## Motivation
+
+Feature gates wrap new cluster functionality, and typically change the rules of
+consensus. As such, a feature gate needs to be supported by a strong majority
+of cluster stake when it is activated, or else it risks partitioning the
+network. The current feature-gate system comprises two steps:
+
+1. An individual key-holder queues a feature gate for activation
+2. The runtime automatically activates the feature on the next epoch boundary
+
+The key-holder is the one who (with the help of the solana-cli) assesses the
+amount of stake that recognizes the feature and decides whether it is safe to
+activate. This is obviously brittle and subject to human error.
+
+This SIMD proposes that the runtime itself replaces the human to assess the
+amount of stake that supports a particular queued feature and only
+automatically activates it when a preset threshold is met. 
+
+## New Terminology
+
+- Feature Gate program: The BPF program that will own all feature accounts.
+- Feature-Request account: The PDA under the Feature Gate program used to track
+features requested for activation.
+- Feature-Queue account: The PDA under the Feature Gate program used to track
+features queued for activation that must have support signaled by nodes.
+
+## Detailed Design
+
+Before this SIMD is published, we will deploy a BPF program to manage feature
+activations, which will allow us to revoke pending feature activations
+(previously not possible).
+The feature activation process will otherwise be unchanged, so no SIMD is
+required.
+This program will be upgradeable; the authority and process through which it
+will be upgraded is outside the scope of this SIMD. Since this SIMD requires an
```

Comment:
Agree with @mvines, we need another SIMD to define this program and interactions with it. In particular it needs to layout:
- What are the instructions to this new program, their inputs, and the side effects. This should probably look like [the system program docs](https://docs.rs/solana-program/1.17.1/solana_program/system_instruction/enum.SystemInstruction.html).
- Who can create new features? I can imagine that anyone can create a new feature account but the clients must recognize it. Maybe clients should vote on if a feature is valid? Maybe stake should vote on the intent to create new features? Lots to consider there. Also consider that allowing anyone to create a feature allows them to potentially slow down the runtime as we would need to process all of these feature accounts (what would you do if there were 500k feature accounts!).
- How are features which are "fully ratified" (i.e. everyone has had them active for a while) reaped? We want to eventually be able to remove feature gates from the code as they can have performance and readability impacts. How will we be able to safely remove them and what are the criteria for removal?
- How are features initialized at genesis? This is really important for testing!

Reply:
Done! #88 #89